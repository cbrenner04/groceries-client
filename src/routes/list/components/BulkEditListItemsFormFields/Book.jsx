import React from 'react';
import PropTypes from 'prop-types';

import { CheckboxField, TextField } from '../../../../components/FormFields';

const Book = ({ author, clearAuthor, handleClearAuthor, handleInput }) => (
  <TextField
    name="author"
    label="Author"
    value={author}
    handleChange={handleInput}
    placeholder="Kurt Vonnagut"
    disabled={clearAuthor}
    child={
      <CheckboxField
        name="clearAuthor"
        label="Clear author"
        handleChange={handleClearAuthor}
        value={clearAuthor}
        classes="ml-1 mt-1"
      />
    }
  />
);

Book.propTypes = {
  author: PropTypes.string.isRequired,
  clearAuthor: PropTypes.bool.isRequired,
  handleClearAuthor: PropTypes.func.isRequired,
  handleInput: PropTypes.func.isRequired,
};

export default Book;
