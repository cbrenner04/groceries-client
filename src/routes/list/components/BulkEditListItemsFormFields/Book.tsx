import React, { type ChangeEventHandler } from 'react';

import { CheckboxField, TextField } from 'components/FormFields';

export interface IBookProps {
  author: string;
  clearAuthor: boolean;
  handleClearAuthor: ChangeEventHandler;
  handleInput: ChangeEventHandler;
}

const Book: React.FC<IBookProps> = (props): React.JSX.Element => (
  <TextField
    name="author"
    label="Author"
    value={props.author}
    handleChange={props.handleInput}
    placeholder="Kurt Vonnagut"
    disabled={props.clearAuthor}
    child={
      <CheckboxField
        name="clearAuthor"
        label="Clear author"
        handleChange={props.handleClearAuthor}
        value={props.clearAuthor}
        classes="ms-1 mt-1"
      />
    }
  />
);

export default Book;
