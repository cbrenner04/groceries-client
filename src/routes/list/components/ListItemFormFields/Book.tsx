import React, { type ChangeEventHandler } from 'react';
import { Row } from 'react-bootstrap';

import { CategoryField, TextField, NumberField, CheckboxField } from 'components/FormFields';

export interface IBookFormFieldsProps {
  author: string;
  title: string;
  purchased: boolean;
  read: boolean;
  editForm: boolean;
  numberInSeries?: number;
  category: string;
  categories: string[];
  inputChangeHandler: ChangeEventHandler;
}

const BookFormFields: React.FC<IBookFormFieldsProps> = (props): React.JSX.Element => (
  <React.Fragment>
    <TextField
      name="author"
      label="Author"
      value={props.author}
      handleChange={props.inputChangeHandler}
      placeholder="Kurt Vonnagut"
    />
    <TextField
      name="title"
      label="Title"
      value={props.title}
      handleChange={props.inputChangeHandler}
      placeholder="Slaughterhouse-Five"
    />
    <NumberField
      name="number_in_series"
      label="Number in series"
      value={props.numberInSeries ?? 0}
      handleChange={props.inputChangeHandler}
    />
    <CategoryField category={props.category} categories={props.categories} handleInput={props.inputChangeHandler} />
    {props.editForm && (
      <Row className="mb-3">
        <CheckboxField
          name="purchased"
          label="Purchased"
          value={props.purchased}
          handleChange={props.inputChangeHandler}
          classes="form-check-inline ms-1"
        />
        <CheckboxField
          name="read"
          label="Read"
          value={props.read}
          handleChange={props.inputChangeHandler}
          classes="form-check-inline"
        />
      </Row>
    )}
  </React.Fragment>
);

export default BookFormFields;
