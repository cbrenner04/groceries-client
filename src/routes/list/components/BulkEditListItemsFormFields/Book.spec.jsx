import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Book from './Book';

function setup(suppliedProps) {
  const user = userEvent.setup();
  const defaultProps = { author: 'foo', clearAuthor: false, handleClearAuthor: jest.fn(), handleInput: jest.fn() };
  const props = { ...defaultProps, ...suppliedProps };
  const { container, findByLabelText, findByRole } = render(<Book {...props} />);

  return { container, findByLabelText, findByRole, props, user };
}

describe('Book', () => {
  it('renders author input enabled when clearAuthor is false', async () => {
    const { container, findByLabelText } = setup({ clearAuthor: false });

    expect(container).toMatchSnapshot();
    expect(await findByLabelText('Author')).toBeEnabled();
  });

  it('renders author input disabled when clearAuthor is true', async () => {
    const { container, findByLabelText } = setup({ clearAuthor: true });

    expect(container).toMatchSnapshot();
    expect(await findByLabelText('Author')).toBeDisabled();
  });

  it('handles change in input and checkbox', async () => {
    const { findByLabelText, findByRole, props, user } = setup({ clearAuthor: false });

    await user.type(await findByLabelText('Author'), 'a');

    expect(props.handleInput).toHaveBeenCalled();

    await user.click(await findByRole('checkbox'));

    expect(props.handleClearAuthor).toHaveBeenCalled();
  });
});
