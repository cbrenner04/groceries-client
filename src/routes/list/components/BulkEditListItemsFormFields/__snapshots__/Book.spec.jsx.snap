// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book renders author input disabled when clearAuthor is true 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="author"
    >
      Author
    </label>
    <input
      class="form-control"
      disabled=""
      id="author"
      name="author"
      placeholder="Kurt Vonnagut"
      type="text"
      value="foo"
    />
    <div
      class="ms-1 mt-1 mb-3"
    >
      <div
        class="form-check"
      >
        <input
          checked=""
          class="form-check-input"
          id="clearAuthor"
          name="clearAuthor"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="clearAuthor"
          title=""
        >
          Clear author
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Book renders author input enabled when clearAuthor is false 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="author"
    >
      Author
    </label>
    <input
      class="form-control"
      id="author"
      name="author"
      placeholder="Kurt Vonnagut"
      type="text"
      value="foo"
    />
    <div
      class="ms-1 mt-1 mb-3"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="clearAuthor"
          name="clearAuthor"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="clearAuthor"
          title=""
        >
          Clear author
        </label>
      </div>
    </div>
  </div>
</div>
`;
